test/test_torch.py

```diff
@@ -98,6 +99,47 @@ def __exit__(self, *args):
         pass
 
 
+def _rebuild_subclass(type_, data, requires_grad, backward_hooks):
+    param = type_(data, requires_grad)
+    # NB: This line exists only for backwards compatibility; the
+    # general expectation is that backward_hooks is an empty
+    # OrderedDict.  See Note [Don't serialize hooks]
+    param._backward_hooks = backward_hooks
+
+    return param
+
+
+class TensorSubclass(torch.Tensor):
+
+    def __new__(cls, data=None, requires_grad=False):
+        if data is None:
+            data = torch.Tensor()
+        self = torch.Tensor._make_subclass(cls, data, requires_grad)
+        return self
+
+    def __deepcopy__(self, memo):
+        if id(self) in memo:
+            return memo[id(self)]
+        else:
+            result = type(self)(self.data.clone(), self.requires_grad)
+            memo[id(self)] = result
+            return result
+
+    def __reduce_ex__(self, proto):
+        # See Note [Don't serialize hooks]
+        return _rebuild_subclass, (self.__class__, self.data, self.requires_grad, OrderedDict())
+
+
+class A(TensorSubclass):
+    pass
+
+class B(TensorSubclass):
+    pass
+
+class C(A, B):
+    pass
```

Comment:
It's really interesting that you added support for complicated class hierarchies, but it's not clear to me what good use-cases for this are. Do you have some good descriptions?

One reason I ask about this, is that I am not certain that the correct thing to do is to unconditionally wrap in the most derived subclass. For example, if I think about someone implementing "named" tensors as a tensor subclass, and then someone else separately implementing "debug" tensors as a tensor subclass, returning either a named tensor or a debug tensor as a result of combining the two is unlikely to be a useful operation. In fact, there isn't really any good thing we can do here, we should probably just error.

Reply:
I don't need class hierarchies personally.  I think I took this directly from how numpy tests overloading numpy methods using `__array_ufunc__`. I have no preference, but I know this way matches numpy.

Reply:
Example:

For class C(A, B), numpy tests that `result = a + c` will call the definition of `C.add()` which inherits from `A.add()` and ignores `B.add()`.

Reply:
I just looked at how I use this. Since there are both Tensors and Parameters, I actually use multiple inheritance and hierarchal inheritance. Here is an example of something that I called TFTensor:

```python
class BaseTensor(th.Tensor): # Used to add new methods to all my tensors at the python level.

class TFTensor(base.BaseTensor): # A Tensor where time and frequency are the last two dimensions.

    def __repr__(self):
        return '%s(..., %s, %s)\n%s' % (TFTensor.__name__, "t", "f", super(TFTensor, self).__repr__())

   # Several other methods are added here
 
class TFParameter(TFTensor, Parameter): # A TFTensor that is used in a neural network.
    pass
```

Note: BaseTensor is just a convenience that I use for adding numpy compatible methods when porting code from numpy to pytorch. I might get rid of it soon.